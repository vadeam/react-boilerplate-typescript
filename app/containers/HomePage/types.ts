import { Record } from 'immutable'

export interface HomeActionTypes {
  CHANGE_USERNAME: string
}

export interface HomeAction extends Action<HomeActionTypes> {
  readonly username: string
}

export interface HomeActionCreators {
  changeUsername: (username: string) => HomeAction
}

export interface HomeState extends InheritedReducerState {
  readonly username: string
}

export const HomeStateRecord = Record<HomeState>(
  {
    username: '',
  },
  'HomeStateRecord',
)
