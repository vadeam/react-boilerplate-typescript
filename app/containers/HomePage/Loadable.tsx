/**
 * Asynchronously loads the component for HomePage
 */

import * as React from 'react'

import LoadingIndicator from 'components/LoadingIndicator'
import loadable from 'utils/loadable'

export default loadable(() => import('./index'), {
  fallback: <LoadingIndicator />,
})
