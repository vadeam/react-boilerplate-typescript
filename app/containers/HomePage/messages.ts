/**
 * HomePage Messages
 * This contains all the text for the HomePage component.
 */
import { defineMessages } from 'react-intl'

export const scope = 'containers.HomePage'

export default defineMessages({
  startProjectHeader: {
    id: `${scope}.start_project.header`,
    defaultMessage: 'Start your next react project in seconds (default message)',
  },
  startProjectMessage: {
    id: `${scope}.start_project.message`,
    defaultMessage:
      'A highly scalable, offline-first foundation with the best DX and a focus on performance and best practices',
  },
  trymeHeader: {
    id: `${scope}.tryme.header`,
    defaultMessage: 'Try me!',
  },
  trymeMessage: {
    id: `${scope}.tryme.message`,
    defaultMessage: 'Show Github repositories by',
  },
  trymeAtPrefix: {
    id: `${scope}.tryme.atPrefix`,
    defaultMessage: '@',
  },
  title: {
    id: `${scope}.title`,
    defaultMessage: 'Home Page',
  },
  frankfurterHeader: {
    id: `${scope}.frankfurter.header`,
    defaultMessage: 'Frankfurter Rates',
  },
  frankfurterLabel: {
    id: `${scope}.frankfurter.label`,
    defaultMessage: 'Base currency',
  },
})
