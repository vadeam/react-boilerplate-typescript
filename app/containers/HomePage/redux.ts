import { createActions, createReducer } from 'reduxsauce'
import { Record, get } from 'immutable'
import { createSelector } from 'reselect'

import { REDUCER_KEY_HOME } from 'constants/reducers'

import { HomeActionCreators, HomeActionTypes, HomeState, HomeStateRecord, HomeAction } from './types'

const prefix = `[${REDUCER_KEY_HOME}] `
const INITIAL_STATE: Record<HomeState> = HomeStateRecord()
const { Types, Creators: actions } = createActions<HomeActionTypes, HomeActionCreators>(
  {
    changeUsername: ['username'],
  },
  { prefix },
)

const selectHome = (state: ApplicationRootState) => state[REDUCER_KEY_HOME] as Record<HomeState>
const selectors = {
  selectHome,
  getUsername: () => createSelector(selectHome, (homeState): string => get(homeState, 'username', '')),
}

const setUsername: ReducerHandler<Record<HomeState>, HomeAction> = (state, { username }) =>
  state.merge({ username: username.replace(/@/gi, '') })

const reducer = createReducer<Record<HomeState>>(INITIAL_STATE, {
  [Types.CHANGE_USERNAME as string]: setUsername,
})

export { Types as types, prefix as namespace, INITIAL_STATE, actions, reducer, selectors }
