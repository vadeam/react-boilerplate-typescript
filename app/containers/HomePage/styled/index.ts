export * from './InputStyled'
export * from './FormStyled'
export * from './SectionStyled'
export * from './DisclaimerStyled'
export * from './CurrencyBlockStyled'
export * from './AtPrefixStyled'
export * from './ContentWrapperStyled'
export * from './FrankfurterLabelStyled'
