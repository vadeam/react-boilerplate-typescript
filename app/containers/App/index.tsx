/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import * as React from 'react'
import { Helmet } from 'react-helmet-async'
import { hot } from 'react-hot-loader/root'
import { Route, Switch } from 'react-router-dom'

import Footer from 'components/Footer'
import Header from 'components/Header'
import FeaturePage from 'containers/FeaturePage/Loadable'
import HomePage from 'containers/HomePage/Loadable'
import NotFoundPage from 'containers/NotFoundPage/Loadable'
import { ThemeProvider } from 'styles/styled-components'
import defaultAppTheme from 'styles/default-theme'
import GlobalStyle from 'global-styles'

import { AppWrapperStyled } from './styled'

function App() {
  return (
    <ThemeProvider theme={defaultAppTheme}>
      <AppWrapperStyled>
        <Helmet titleTemplate="%s | React Boilerplate" defaultTitle="React Boilerplate">
          <meta name="description" content="React Boilerplate application" />
        </Helmet>
        <Header />
        <Switch>
          <Route exact path="/" component={HomePage} />
          <Route path="/features" component={FeaturePage} />
          <Route component={NotFoundPage} />
        </Switch>
        <Footer />
        <GlobalStyle />
      </AppWrapperStyled>
    </ThemeProvider>
  )
}

export default hot(App)
