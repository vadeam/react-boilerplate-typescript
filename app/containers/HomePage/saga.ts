/**
 * Gets the repositories of the user from Github
 */

import { call, put, select, takeLatest } from 'redux-saga/effects'

import { types as AppActionTypes, actions as AppActions } from 'containers/App/redux'
import fetcher from 'utils/fetcher'

import { selectors } from './redux'

/**
 * Github repos request/response handler
 */
export function* getRepos() {
  const username = yield select(selectors.getUsername())

  try {
    const repos = yield call(fetcher.getUserGithubRepos, username)
    yield put(AppActions.reposLoaded(repos, username))
  } catch (err) {
    yield put(AppActions.repoLoadingError(err))
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* githubData() {
  yield takeLatest(AppActionTypes.LOAD_REPOS, getRepos)
}
