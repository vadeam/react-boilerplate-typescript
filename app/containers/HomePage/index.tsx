/**
 * HomePage
 * This is the first thing users see of our App, at the '/' route
 */

import React, { useCallback, useState } from 'react'
import { Helmet } from 'react-helmet-async'
import { FormattedMessage, useIntl } from 'react-intl'
import { useDispatch, useSelector } from 'react-redux'
import { createStructuredSelector } from 'reselect'
import useSWR from 'swr'

import H2 from 'components/H2'
import ReposList from 'components/ReposList'
import RatesList from 'components/RatesList'
import Toggle from 'components/Toggle'
import { selectors as appSelectors, actions as appActions } from 'containers/App/redux'
import { useInjectReducer, useInjectSaga } from 'utils/redux-injectors'
import { REDUCER_KEY_HOME } from 'constants/reducers'
import { DEFAULT_BASE_CURRENCY } from 'constants/franfurter'

import fetcher from '../../utils/fetcher'

import {
  AtPrefixStyled,
  DisclaimerStyled,
  CurrencyBlockStyled,
  FormStyled,
  InputStyled,
  SectionStyled,
  ContentWrapperStyled,
  FrankfurterLabelStyled,
} from './styled'
import messages from './messages'
import { actions, reducer, selectors } from './redux'
import saga from './saga'

const stateSelector = createStructuredSelector({
  repos: appSelectors.makeSelectRepos(),
  username: selectors.getUsername(),
  isFetching: appSelectors.makeSelectFetching(),
  error: appSelectors.makeSelectError(),
})

export default function HomePage() {
  const { formatMessage } = useIntl()
  const { username, repos, isFetching, error } = useSelector(stateSelector)
  const dispatch = useDispatch()
  // Not gonna declare event types here. No need. any is fine
  const onChangeUsername = useCallback((evt: any) => dispatch(actions.changeUsername(evt.target.value)), [dispatch])
  const onSubmitForm = useCallback(
    (evt: any) => {
      if (evt !== undefined && evt.preventDefault) {
        evt.preventDefault()
      }
      if (!username) {
        dispatch(appActions.reposLoaded(null, ''))
      } else {
        dispatch(appActions.loadRepos())
      }
    },
    [username, dispatch],
  )

  const [baseCurrency, setBaseCurrency] = useState<string>(DEFAULT_BASE_CURRENCY)
  const onBaseCurrencyToggle = useCallback((evt) => setBaseCurrency(evt.target.value), [])
  const { data: currenciesResponse } = useSWR<{ [k: string]: string }>('/currencies', fetcher.getFranfurterData)

  useInjectReducer({ key: REDUCER_KEY_HOME, reducer })
  useInjectSaga({ key: REDUCER_KEY_HOME, saga })

  return (
    <article>
      <Helmet>
        <title>{formatMessage({ ...messages.title })}</title>
        <meta name="description" content="A React Boilerplate application homepage" />
      </Helmet>
      <div>
        <DisclaimerStyled>
          <H2>
            <FormattedMessage {...messages.startProjectHeader} />
          </H2>
          <p>
            <FormattedMessage {...messages.startProjectMessage} />
          </p>
        </DisclaimerStyled>
        <ContentWrapperStyled>
          <SectionStyled width="70%">
            <H2>
              <FormattedMessage {...messages.trymeHeader} />
            </H2>
            <FormStyled onSubmit={onSubmitForm}>
              <label htmlFor="username">
                <FormattedMessage {...messages.trymeMessage} />
                <AtPrefixStyled>
                  <FormattedMessage {...messages.trymeAtPrefix} />
                </AtPrefixStyled>
                <InputStyled
                  id="username"
                  type="text"
                  placeholder="mxstbr"
                  value={username}
                  onChange={onChangeUsername}
                />
              </label>
            </FormStyled>
            <ReposList {...{ isFetching, error, repos }} />
          </SectionStyled>
          <SectionStyled width="30%">
            <H2>
              <FormattedMessage {...messages.frankfurterHeader} />
            </H2>
            <CurrencyBlockStyled>
              <FrankfurterLabelStyled>
                <FormattedMessage {...messages.frankfurterLabel} />
                :&nbsp;&nbsp;&nbsp;
              </FrankfurterLabelStyled>
              <Toggle
                value={baseCurrency}
                values={Object.keys(currenciesResponse ?? {})}
                onToggle={onBaseCurrencyToggle}
              />
            </CurrencyBlockStyled>
            <RatesList baseCurrency={baseCurrency} />
          </SectionStyled>
        </ContentWrapperStyled>
      </div>
    </article>
  )
}
