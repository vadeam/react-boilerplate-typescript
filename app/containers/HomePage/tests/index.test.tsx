/**
 * Test the HomePage
 */

import { render, cleanup, fireEvent } from '@testing-library/react'
import React from 'react'
import { HelmetProvider } from 'react-helmet-async'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'

import { actions } from 'containers/App/redux'

import configureStore from '../../../configureStore'
import history from '../../../utils/history'
import HomePage from '../index'
import { INITIAL_STATE } from '../redux'
import { DEFAULT_LOCALE } from '../../../constants/locales'

const renderHomePage = (store) =>
  render(
    <Provider store={store}>
      <IntlProvider locale={DEFAULT_LOCALE}>
        <HelmetProvider>
          <HomePage />
        </HelmetProvider>
      </IntlProvider>
    </Provider>,
  )

describe('<HomePage />', () => {
  let store
  const mockedLoadRepos = jest.spyOn(actions, 'loadRepos').mockImplementation(() => ({ type: '' } as any))

  beforeEach(() => {
    store = configureStore({}, history)
    mockedLoadRepos.mockClear()
  })

  afterEach(cleanup)

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = renderHomePage(store)
    expect(firstChild).toMatchSnapshot()
  })

  it('shouldn`t fetch repos on mount (if username is empty)', () => {
    renderHomePage(store)
    expect(INITIAL_STATE.toJS().username).toBe('')
    expect(mockedLoadRepos).not.toHaveBeenCalled()
  })

  it('shouldn`t fetch repos if the form is submitted when the username is empty', () => {
    const { container } = renderHomePage(store)

    const form = container.querySelector('form')!
    fireEvent.submit(form)

    expect(mockedLoadRepos).not.toHaveBeenCalled()
  })

  it('should fetch repos if the form is submitted when the username isn`t empty', () => {
    const { container } = renderHomePage(store)
    const input = container.querySelector('input')!

    fireEvent.change(input, { target: { value: '' } })
    expect(mockedLoadRepos).not.toHaveBeenCalled()

    fireEvent.change(input, { target: { value: 'julienben' } })
    expect(mockedLoadRepos).not.toHaveBeenCalled()

    const form = container.querySelector('form')!

    fireEvent.submit(form)
    expect(mockedLoadRepos).toHaveBeenCalledTimes(1)
  })
})
