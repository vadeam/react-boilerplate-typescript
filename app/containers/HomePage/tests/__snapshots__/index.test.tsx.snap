// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<article>
  <div>
    <section
      class="sc-crzoAE sc-dIsUp ertcJF gsOIfo"
    >
      <h2
        class="sc-bdnxRM cIIgpD"
      >
        <span>
          Start your next react project in seconds
        </span>
      </h2>
      <p>
        <span>
          A highly scalable, offline-first foundation with the best DX and a focus on performance and best practices
        </span>
      </p>
    </section>
    <section
      class="sc-crzoAE ertcJF"
    >
      <h2
        class="sc-bdnxRM cIIgpD"
      >
        <span>
          Try me!
        </span>
      </h2>
      <form
        class="sc-bqGGPW iAPa-ds"
      >
        <label
          for="username"
        >
          <span>
            Show Github repositories by
          </span>
          <span
            class="sc-iqAclL fJPgGE"
          >
            <span>
              @
            </span>
          </span>
          <input
            class="sc-ksluID gjUEoZ"
            id="username"
            placeholder="mxstbr"
            type="text"
            value=""
          />
        </label>
      </form>
      <div
        class="sc-dlnjwi gHUNA-D"
      >
        <ul
          class="sc-gtsrHT iMOCkE"
        />
      </div>
    </section>
  </div>
</article>
`;
